components/raftstore/src/store/peer_storage.rs

```diff
@@ -436,16 +444,41 @@ where
     /// Gets a snapshot. Returns `SnapshotTemporarilyUnavailable` if there is no
     /// unavailable snapshot.
     pub fn snapshot(&self, request_index: u64, to: u64) -> raft::Result<Snapshot> {
+        if self.peer.as_ref().unwrap().is_witness {
+            // witness could be the leader for a while, do not generate snapshot now
+            return Err(raft::Error::Store(
+                raft::StorageError::SnapshotTemporarilyUnavailable,
+            ));
+        }
+
+        let for_witness = find_peer_by_id(&self.region, to).unwrap().is_witness;
+        if for_witness {
+            // generate an empty snapshot for witness directly
+            let mut snapshot = Snapshot::default();
```

Comment:
Can this be extracted as a function? It can be reused for normal snapshot and v2.